### ISSUE TITLE
curl request for latest firmware

### ISSUE THREAD

**Issue by parkerlreed** [luigimannoni/stadia-controller-flasher#2]
Figured this would be good to archive for if/when this goes down.

The request has the "current" slot A or B and gives you back the opposite slot. So yes definitely supports two firmware.

```
(deck@steamdeck src)$ curl -s 'https://update.googleapis.com/service/update2' --data-binary '@-' -H 'Content-Type: text/xml' <<'EOF' | xmllint --format -
<?xml version="1.0" encoding="UTF-8"?>
<gupdate protocol="3.0" xmlns="http://www.google.com/update2/request">
<app appid="{50940242-7306-4f20-80f1-dde7c638f235}" track="stable-channel" board="gotham-pvt" image="b">
    <updatecheck />
</app>
</gupdate>
EOF

<?xml version="1.0" encoding="UTF-8"?>
<response protocol="3.0" server="prod">
  <daystart elapsed_days="5863" elapsed_seconds="57611"/>
  <app appid="{50940242-7306-4f20-80f1-dde7c638f235}" cohort="1::" cohortname="" status="ok">
    <updatecheck status="ok">
      <urls>
        <url codebase="http://edgedl.me.gvt1.com/edgedl/googletv-eureka/stable-channel/"/>
        <url codebase="http://redirector.gvt1.com/edgedl/googletv-eureka/stable-channel/"/>
      </urls>
      <manifest version="282115">
        <actions>
          <action event="install" run="ota.282115.stable-channel.gotham-pvt-a.214f6fe20f7b993046c77a9245c5805b056882dc.bin"/>
          <action DisplayVersion="282115" DoFDR="false" DownloadRate="0" Foreground="true" RequireCharging="false" UpdateTiming="1" event="postinstall"/>
        </actions>
        <packages>
          <package fp="1.269bb21160ea6d1366e71bb9c9b148e2782ea967b81c7b6ff038e7d49b61e00f" hash="IU9v4g97mTBGx3qSRcWAWwVogtw=" hash_sha256="269bb21160ea6d1366e71bb9c9b148e2782ea967b81c7b6ff038e7d49b61e00f" name="ota.282115.stable-channel.gotham-pvt-a.214f6fe20f7b993046c77a9245c5805b056882dc.bin" required="true" size="4163944"/>
        </packages>
      </manifest>
    </updatecheck>
  </app>
</response>

(deck@steamdeck src)$ curl -s 'https://update.googleapis.com/service/update2' --data-binary '@-' -H 'Content-Type: text/xml' <<'EOF' | xmllint --format -
<?xml version="1.0" encoding="UTF-8"?>
<gupdate protocol="3.0" xmlns="http://www.google.com/update2/request">
<app appid="{50940242-7306-4f20-80f1-dde7c638f235}" track="stable-channel" board="gotham-pvt" image="a">
    <updatecheck />
</app>
</gupdate>
EOF

<?xml version="1.0" encoding="UTF-8"?>
<response protocol="3.0" server="prod">
  <daystart elapsed_days="5863" elapsed_seconds="57617"/>
  <app appid="{50940242-7306-4f20-80f1-dde7c638f235}" cohort="1::" cohortname="" status="ok">
    <updatecheck status="ok">
      <urls>
        <url codebase="http://edgedl.me.gvt1.com/edgedl/googletv-eureka/stable-channel/"/>
        <url codebase="http://redirector.gvt1.com/edgedl/googletv-eureka/stable-channel/"/>
      </urls>
      <manifest version="282115">
        <actions>
          <action event="install" run="ota.282115.stable-channel.gotham-pvt-b.b252d0d5a16538621abfb09609746cabc0f33292.bin"/>
          <action DisplayVersion="282115" DoFDR="false" DownloadRate="0" Foreground="true" RequireCharging="false" UpdateTiming="1" event="postinstall"/>
        </actions>
        <packages>
          <package fp="1.1cbc35203ff83a5244a8a9b2590938983f162259dfdb6235321d60e9356ce44e" hash="slLQ1aFlOGIav7CWCXRsq8DzMpI=" hash_sha256="1cbc35203ff83a5244a8a9b2590938983f162259dfdb6235321d60e9356ce44e" name="ota.282115.stable-channel.gotham-pvt-b.b252d0d5a16538621abfb09609746cabc0f33292.bin" required="true" size="4163944"/>
        </packages>
      </manifest>
    </updatecheck>
  </app>
</response>
```

**Comment by parkerlreed** [comment]
Also I can't take full credit for the URL you got or this. You can thank @tchebb !

I'm just happy to spread the info.

**Comment by parkerlreed** [comment]
Here's the B slot for latest release building from that request

http://edgedl.me.gvt1.com/edgedl/googletv-eureka/stable-channel/ota.282115.stable-channel.gotham-pvt-b.b252d0d5a16538621abfb09609746cabc0f33292.bin

**Comment by parkerlreed** [comment]
I'm currently figuring out what change is needed to grab the other dogfood.

**Comment by parkerlreed** [comment]
Corrected the URL above. I forgot we already had A latest.

**Comment by luigimannoni** [comment]
Thanks, that's literally a gold mine and probably it's quite easy to stick the actual firmware URLs onto the current web app.

Just out of curiosity, this is the latest official consumer build, but is there a way to actually request the unreleased build 316328?


On another note, I do imagine their endpoints will cease working in the future and keeping the firmwares in the repo will probably anger some people over at Google. 

So ultimately I guess it should be a "Bring/backup-your-own-firmware" type of utility

**Comment by parkerlreed** [comment]
I mean at a certain point it's hardware they don't care about anymore. 

And yes that's what I meant above trying to figure out how to modify the curl request to pull in the unreleased. I've tried changing stable to a few things but haven't got it yet. I asked the guy I mentioned above who created it and just waiting for a response.

**Comment by parkerlreed** [comment]
I totally forgot that over the years I've been trying to capture every update that I saw. Here's a few more older ones

More dogfood builds
http://edgedl.me.gvt1.com/edgedl/googletv-eureka/stable-channel/ota.269117.stable-channel.gotham-pvt-b.d98aee34c392bf64f198f3f5a24018961f72d461.bin
http://edgedl.me.gvt1.com/edgedl/googletv-eureka/stable-channel/ota.269117.stable-channel.gotham-pvt-a.1131ce3771b3da5429417d4d1935b7d2b9943fad.bin

Older stable release http://edgedl.me.gvt1.com/edgedl/googletv-eureka/stable-channel/ota.259108.stable-channel.gotham-pvt-b.0f4b0353c57c9e62f7e4d5a8a847853f7381aa6b.bin

**Comment by kwuite** [comment]
@luigimannoni which images need to be pulled here? Just the latest firmware and that's it? Does it need to be configurable and if so, how will this be handled by the end user?

Will you be adding the processing of flags via the user interface?  
- '?device_type=' 
- '?firmware_source' - for alternative sources
  - local and remote?
- '?persistent' - for cache?

How will this affect your github pages? Have you taken this into consideration (like will you pull them during the github action process)?

**Comment by luigimannoni** [comment]
I'd rather prefer if the flasher downloads the images automatically from the google servers and flash whatever the user selects on the interface as I am certain if firmwares are not removed a DMCA will hit the repo some time in the future and everything goes down with it.

I'm open to suggestions but I think there should be a list of sources on code and the ability to select a file. Lets assume if [someone dumped their firmware](https://github.com/GaryOderNichts/StadiaController) and want to reflash or if they have downloaded a file already, then they should be able to choose whatever they want.
Also it is possible in 3/4 months google will remove firmwares from their servers so in the long term makes sense to select the firmware file via interface.

So in short, both deploy and binary packages don't need to contain images, I understand that this means the interface needs to be rebuilt from scratch to add new functionality as customisation is difficult with the minified code in the current app, see #6 


I'll have think probably this evening and give it a try to migrate all the app into a re-usable framework

**Comment by kwuite** [comment]
The DMCA is very much a threat indeed and Google will problably remove the firmware as they fired almost everyone involved in the Stadia project so I don't see anyone maintaining that logic.

And your requirements make perfect sense:
- flasher downloads the images automatically
- list of sources on code and the ability to select a file
- deploy and binary packages "should not" need to contain images
- local file firmware capability
